### ISSUE TITLE
Glyph selection without positioning individual glyphs

### ISSUE THREAD

**Issue by Kethku** [dfrg/swash#9]
This may very well be out of scope for the library.

I'm very interested in swash as a font parsing engine for https://github.com/Kethku/neovide which is a gui for Neovim written in rust. It uses skia for rendering, but the built in harfbuzz bindings in skia leave much to be desired. Effectively Neovide is just a fancy terminal renderer which draws characters in a grid. However we use harfbuzz to select which glyphs to render for the text. This lets us combine certain glyphs into ligatures.

Currently we rely on harfbuzz to also position the glyphs, but this leads to grid positioning problems and I would like to eventually position the glyphs myself based on user config and some basic heuristics that assume fonts are monospace. The problems occur when font fallback produces a font which has different spacing from the font used to align the grid. So you get misaligned text. However if I position glyphs manually, this problem goes away.

So what would be awesome is a way to just use the glyph selection portion of the shaper rather than going through the entire shaping process. I'm not interested in the glyph positions because I'm going to place the glyphs myself on the monospace grid. I'm only interested in when a substitution is made which turns 1 or more glyphs into a single ligature, or similar such swaps. I'm not familiar with all of the types of substitutions that can be made, but I'm wondering if such a thing would be possible.

I've looked into doing this with harfbuzz, and I think it would also be possible to do it with it, but to my knowledge it would require using the clustering mechanism which I find VERY confusing. Maybe this use case could be better covered in swash? No biggy if not, just figured I'd ask.

As an aside, I would LOVE to get rid of the skia dependency all together as building it is hard and being pure rust would be badass. I've looked at pathfinder before, but was confused by the various levels at which I could integrate and eventually just went with skia as the easier thing to get myself up and running faster. You mention pathfinder in the readme as a possible renderer for swash's output. Have you done such a thing? Is there an example of such an integration? I'd be interested

**Comment by Kethku** [comment]
Side note: I just swapped to swash over from rustybuzz in my development branch of neovide. Initial tests seem to indicate that it is WAY faster than my usage of rustybuzz. I am very impressed.

**Comment by dfrg** [comment]
Hi Keith,

So sorry for the very late response-- I've been otherwise occupied with some personal issues but will be devoting a large portion of my time to development of this and related projects (fount and parley) for the foreseeable future.

It would potentially be possible to disable the glyph positioning features and just use the substitutions. The problem here would be proper positioning of marks. There is a late pass in the shaper that adjusts mark offsets to account for the advance of their associated base glyph. This allows very simple sequential positioning of glyphs during layout without having to make any further computations for attachments. HarfBuzz does something similar as it produces equivalent offsets for diacritics and other marks. There are also some gnarly situations where you can have marks attached to specific components of a ligature, so the resulting clusters can be fairly complex. 

Detecting ligatures produced by the shaper is a much simpler problem. The `GlyphCluster` struct has a predicate method `is_ligature` that reports it directly. The `components` field will contain the source ranges of each character that contributed to the ligature. Note that it's still possible for a ligature cluster to contain multiple glyphs due to combining marks or through a multiple substitution lookup in GSUB. Unfortunately (or fortunately, depending on your perspective), the substitution tables in a font can be arbitrarily complex and a general purpose shaper can't make any assumptions even if they might seem reasonable.

I have spent some time looking through issues that terminal programs face when dealing with complex text and my take away was that the problem is interesting, but also very difficult. I'm not sure that an ideal solution, exists particularly when you want to support syllabic scripts like those in the Indic family as those don't map in any meaningful way to a character-per-cell layout. 

If you have some specific examples of the text you'd like to render and the output you expect from the shaper, it might help me to better understand the problem in which case I can propose an alternative strategy or perhaps modifications to the shaper.

Re: performance- very glad to see my time buried in the profiler has been of use.

**Comment by dfrg** [comment]
Oops. Forgot to comment on rendering. 

I haven't personally attempted any integration with pathfinder or lyon. They're both very nice vector GPU projects with interesting implementations, but they also suffer from shortcomings that, in my opinion, limit their usefulness for production quality rendering.

The demo I released simply uses my zeno crate for rasterization and contains a very naïve glyph/image cache and OpenGL compositor.

Unfortunately, the Rust ecosystem is currently lacking a high quality, high performance solution for full composition. The closest thing is probably WebRender, but that's also quite a beast. I'm _very_ bullish on Raph Levien's piet-gpu (https://github.com/linebender/piet-gpu) though it's still cooking. The work I'm doing at the moment revolves around integrating this crate with the piet ecosystem (and eventually, piet-gpu) and I'm hopeful that the combination will become the go-to solution for text layout, rendering and composition in the future. For the time being, I'm afraid we're stuck with limited, ad hoc renderers or non-Rust libraries.

**Comment by Kethku** [comment]
> Unfortunately, the Rust ecosystem is currently lacking a high quality, high performance solution for full composition. The closest thing is probably WebRender, but that's also quite a beast. I'm _very_ bullish on Raph Levien's piet-gpu (https://github.com/linebender/piet-gpu) though it's still cooking. The work I'm doing at the moment revolves around integrating this crate with the piet ecosystem (and eventually, piet-gpu) and I'm hopeful that the combination will become the go-to solution for text layout, rendering and composition in the future. For the time being, I'm afraid we're stuck with limited, ad hoc renderers or non-Rust libraries.

This has been our experience as well. We use skia today not because its ideal (it introduces a large build and complexity burden) but because the alternatives have either been unstable, or require other things such as vulkan. Our initial system was built using vulkan as the graphics backend and skia for a high level api, but recently we had to swap to an opengl backend because many linux systems don't play nice with vulkan yet. I'm very excited for piet-gpu, but I'm nervous that it will have similar drawbacks.

**Comment by Kethku** [comment]
> I have spent some time looking through issues that terminal programs face when dealing with complex text and my take away was that the problem is interesting, but also very difficult. I'm not sure that an ideal solution exists, particularly when you want to support syllabic scripts like those in the Indic family as those don't map in any meaningful way to a character-per-cell layout.

I will do some thinking about this and get back to you. Part of the issue is that I don't know enough about the font rendering problem to know what the "correct" solution is for rendering text in a grid. So far my approach with Neovide has been to do something that works how I expect it to for my particular language/ligature/emoji needs and then respond when somebody shouts that its broken. 

Currently we have what amounts to a nasty hack https://github.com/Kethku/neovide/blob/main/src/renderer/fonts/caching_shaper.rs#L104 which uses the unicode graphemes index (what neovim uses to layout glyphs on a grid) as the userdata. Then go through a sorta manual cluster/font selection process as you laid out in the docs to find fonts. Then when we render, the position is calculated using the userdata. I suspect this won't work for nearly every font, but at least got me off the ground with swash.

What you highlight regarding marks and syllabic scripts is interesting. I will have to think about this as well. I think the only way out for my project is to just state up front what types of fonts are supported. We already do this implicitly by assuming fonts are monospace. Extending that constraint a bit may let me do things in the special case which wouldn't be valid for the universe of all possible fonts.

Thanks so much for sharing your insight! And for swash itself. Its been a breath of fresh air compared to the other shaping options I've been fiddling with.

**Comment by dfrg** [comment]
Just wanted to follow up on this a bit. I really appreciate you making use of this crate in neovide! Those animations are sharp and perfectly executed. Makes me wish I could tolerate using a modal editor but I was unfortunately soured on the prospect during my earliest days of programming when I was forced to use something that was a pared down imitation of _ed_.

I took a peek at your initial integration and it was very helpful in exposing what I feel are some real inadequacies in the shaper API when dealing with multiple fonts. My own integration in the swash demo is an absolute nightmare. In response, I wrote another interface to the shaper that encapsulates the process of breaking runs based on a user provided font selector backed by a trait. This hasn't been push to crates.io yet because I'm not completely sold on the API, but it is in the repo and I've made use of it in the layout crate that I'm currently working on: https://github.com/dfrg/parley/blob/master/src/builder.rs#L148

When I settle on the API and do a release, I'd be happy to submit a PR to neovide and work with you on getting it integrated. It should buy you some more performance and eliminate all the complications and allocations around accumulating clusters and breaking runs.
